docusaurus/docs/tutorials/build-a-streaming-data-source-plugin.md

```diff
@@ -15,143 +14,286 @@ keywords:
 
-1. Modify the signature of the `query` method to return an `Observable` from the `rxjs` package. Make sure you remove the `async` keyword.
+- **`tickInterval`** - the interval for generating the data
+- **`upperLimit`** - the upper limit for the random numbers
+- **`lowerLimit`** - the lower limit for the random numbers
 
-   ```ts
-   import { Observable } from 'rxjs';
-   ```
+To accomplish this, follow these steps.
 
-   ```ts
-   query(options: DataQueryRequest<MyQuery>): Observable<DataQueryResponse> {
-     // ...
-   }
-   ```
+1. Replace the `QueryEditor` function in `/src/components/QueryEditor.tsx` with the following code:
 
-1. Create an `Observable` instance for each query, and then combine them all using the `merge` function from the `rxjs` package.
+```
+export function QueryEditor({ query, onChange, onRunQuery }: Props) {
+  const onLowerLimitChange = (event: ChangeEvent<HTMLInputElement>) => {
+    onChange({ ...query, lowerLimit: event.target.valueAsNumber });
+  };
+
+  const onUpperLimitChange = (event: ChangeEvent<HTMLInputElement>) => {
+    onChange({ ...query, upperLimit: event.target.valueAsNumber });
+  };
+
+  const onTickIntervalChange = (event: ChangeEvent<HTMLInputElement>) => {
+    onChange({ ...query, tickInterval: event.target.valueAsNumber });
+  };
+
+  const { upperLimit, lowerLimit, tickInterval } = query;
+
+  return (
+    <div className="gf-form">
+      <InlineField label="Lower Limit" labelWidth={16} tooltip="Random numbers lower limit">
+        <Input onChange={onLowerLimitChange} onBlur={onRunQuery} value={lowerLimit || ''} type="number" />
+      </InlineField>
+      <InlineField label="Upper Limit" labelWidth={16} tooltip="Random numbers upper limit">
+        <Input onChange={onUpperLimitChange} onBlur={onRunQuery} value={upperLimit || ''} type="number" />
+      </InlineField>
+      <InlineField label="Tick interval" labelWidth={16} tooltip="Server tick interval">
+        <Input onChange={onTickIntervalChange} onBlur={onRunQuery} value={tickInterval || ''} type="number" />
+      </InlineField>
+    </div>
+  );
+}
+```
 
-   ```ts
-   import { Observable, merge } from 'rxjs';
-   ```
+1. In the `src/datasource.ts` file, indicate a query through a stream channel. This file is where the query executed by the frontend part of your plugin is made. Add the following method to the `DataSource` class:
 
-   ```ts
-   const observables = options.targets.map((target) => {
-     return new Observable<DataQueryResponse>((subscriber) => {
-       // ...
-     });
-   });
+```
+  query(request: DataQueryRequest<MyQuery>): Observable<DataQueryResponse> {
+    const observables = request.targets.map((target, index) => {
+      const query = defaults(target, DEFAULT_QUERY);
```

Comment:
I think we'd want devs to use the `getDefaultQuery` method rather than setting defaults like this, so you can remove this line.

Reply:
@sunker - To be clear, are you saying that the only change needed here is to remove line 102, or something else?

Reply:
yes - but also change line 101 from 
`const observables = request.targets.map((target, index) => {`
to
`const observables = request.targets.map((query, index) => {`

Reply:
done